erigon-lib/downloader/mdbx_piece_completion.go

```diff
@@ -84,32 +103,83 @@ func (m mdbxPieceCompletion) Set(pk metainfo.PieceKey, b bool) error {
 	//  1K fsyncs/2minutes it's quite expensive, but even on cloud (high latency) drive it allow download 100mb/s
 	//  and Erigon doesn't do anything when downloading snapshots
 	if b {
-		tx, err = m.db.BeginRwNosync(context.Background())
-		if err != nil {
-			return err
-		}
-	} else {
-		tx, err = m.db.BeginRw(context.Background())
-		if err != nil {
-			return err
+		completed, ok := m.completed[pk.InfoHash]
+
+		if !ok {
+			completed = &roaring.Bitmap{}
+			m.completed[pk.InfoHash] = completed
 		}
+
+		completed.Add(uint32(pk.Index))
+
+		return nil
+	}
+
+	tx, err = m.db.BeginRw(context.Background())
+	if err != nil {
+		return err
 	}
+
 	defer tx.Rollback()
 
+	err = putCompletion(tx, pk.InfoHash, uint32(pk.Index), b)
+
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func putCompletion(tx kv.RwTx, infoHash infohash.T, index uint32, c bool) error {
 	var key [infohash.Size + 4]byte
-	copy(key[:], pk.InfoHash[:])
-	binary.BigEndian.PutUint32(key[infohash.Size:], uint32(pk.Index))
+	copy(key[:], infoHash[:])
+	binary.BigEndian.PutUint32(key[infohash.Size:], index)
 
 	v := []byte(incomplete)
-	if b {
+	if c {
 		v = []byte(complete)
 	}
-	err = tx.Put(kv.BittorrentCompletion, key[:], v)
+	//fmt.Println("PUT", infoHash, index, c)
+	return tx.Put(kv.BittorrentCompletion, key[:], v)
+}
+
+func (m *mdbxPieceCompletion) Flushed(infoHash infohash.T, flushed *roaring.Bitmap) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	tx, err := m.db.BeginRwNosync(context.Background())
```

Comment:
please use `BeginRw` instead of `BeginRwNosync`. otherwise you can loose last committed transaction on power-off.

Reply:
Done